Add SelectionModel change information

grokys opened the pull request:
## What does the pull request do?

`SelectionModel` as ported from WinUI (#3469) has no information about what changed in a `SelectionChanged` event. This adds that information along with unit tests.

WinUI issue: https://github.com/microsoft/microsoft-ui-xaml/issues/1645

Part of #3469

grokys committed:
Add `IndexRange` list add/remove methods.

Add or remove index ranges from a list of index ranges, merging and splitting ranges as required.

grokys committed:
Added SelectionModel changed args.

`SelectionModel` as ported from WinUI has no information about what changed in a `SelectionChanged` event. This adds that information along with unit tests.

ranjeshj commented on src/Avalonia.Controls/SelectionNode.cs:
```diff
@@ -276,12 +283,50 @@ public List<int> SelectedIndices
+
+                for (var i = 0; i < _childrenNodes.Count; ++i)
+                {
+                    var child = _childrenNodes[i];
```
nit: could use foreach

kekekeks replied:
it's iterating the list in reverse order and is removing entries, so, no, it couldn't.

ranjeshj replied:
Hmm.. how so ? the loop is from 0 to _chlidrennodes.Count...

kekekeks replied:
Whoops, got confused by another loop.

grokys replied:
Yep, you're right. Got too eager with avoiding allocations there I think (enumerating over `IList<>` allocates, but this is a `List<>` so we should be ok).

ranjeshj commented on src/Avalonia.Controls/IndexRange.cs:
```diff
@@ -3,12 +3,17 @@
-    internal readonly struct IndexRange
+    internal readonly struct IndexRange : IEquatable<IndexRange>
     {
+        private static readonly IndexRange s_invalid = new IndexRange(int.MinValue, int.MinValue);
```
>s_invalid [](start = 43, length = 9)

Curious why this is needed? Can we just not have that index Range instead ? Or is just the initial state ?

grokys replied:
Sorry, don't understand, could we not just have what index Range?

ranjeshj replied:
is s_invalid needed ? when is a range invalid ? An empty range makes sense, but what case does a range become invalid ?

grokys replied:
Got you. Yeah, it's only used internally as a [sentinal value in `Add()`](https://github.com/AvaloniaUI/Avalonia/blob/88794773229e5c9fe93f29672c41b9cbac601a89/src/Avalonia.Controls/IndexRange.cs#L97) to indicate that we're finished adding the range. Adding a separate bool flag would work as well, but I figured that we have a struct with 8,589,934,590 possible values so I might as well use one of those values to indicate that the loop is finished... I assumed that it's unlikely that a range of `[-minvalue..-minvalue]` would be used in a real situation.

ranjeshj commented on src/Avalonia.Controls/SelectionModel.cs:
```diff
@@ -452,9 +489,15 @@ private void RaisePropertyChanged(string propertyName)
+            IEnumerable<object>? removedItems)
         {
-            OnSelectionChanged();
+            var e = new SelectionModelSelectionChangedEventArgs(
```
>new SelectionModelSelectionChangedEventArg [](start = 20, length = 42)

This was cached to avoid repeated allocations. Not sure if that is a concern in this project.

grokys replied:
Yeah I removed the caching because caching could cause a problem if a `SelectionChanged` event handler made a change to the selection itself. However thinking about this some more, because the added/removed items are lazily-evaluated, this could still break...

ranjeshj commented on src/Avalonia.Controls/SelectionNode.cs:
```diff
@@ -28,6 +29,7 @@ internal class SelectionNode : IDisposable
         private readonly SelectionNode? _parent;
         private readonly List<IndexRange> _selected = new List<IndexRange>();
         private readonly List<int> _selectedIndicesCached = new List<int>();
+        private SelectionModelChangeSet? _changes;
```
nit: I wonder if this would be a little more simplified if these were refactored in a separate class by themselves. There is some duplication between _selected and keeping track of the changeset.

grokys replied:
Hmm, could you explain more? I'm struggling to see the duplication.

ranjeshj commented on src/Avalonia.Controls/SelectionModel.cs:
```diff
@@ -691,8 +704,55 @@ private void SelectRangeImpl(IndexPath start, IndexPath end, bool select)
+        {
+            static IEnumerable<T>? Concat<T>(IEnumerable<T>? a, IEnumerable<T> b)
+            {
+                return a == null ? b : a.Concat(b);
```
Not sure how this works, but would we end up enumerating the items even if the app does not request for any of the properties ?

grokys replied:
The items should only be enumerated if the user requests one of the properties as `SelectionModelChangeSet` uses `yield`: https://github.com/AvaloniaUI/Avalonia/blob/88794773229e5c9fe93f29672c41b9cbac601a89/src/Avalonia.Controls/SelectionModelChangeSet.cs#L112-L126

As mentioned before however, this could cause a problem if a selection change is made in the `SelectionChanged` handler... We may need to create a new `SelectionModelChangeSet` object for each operation.

grokys replied:
I've refactored this class, it still only creates the collections on demand but it does actually materialize a collection with all indexes/items. I intend to update it to use the `SelectedItems` collection so that the index/items collections are never actually materialized.

ranjeshj commented on src/Avalonia.Controls/IndexRange.cs:
```diff
@@ -54,6 +57,164 @@ public bool Split(int splitIndex, out IndexRange before, out IndexRange after)
+        public static bool operator ==(IndexRange left, IndexRange right) => left.Equals(right);
+        public static bool operator !=(IndexRange left, IndexRange right) => !(left == right);
+
+        public static int Add(
```
Nice refactor to move these here.

ranjeshj commented:
>                     ClearSelection();

Hmm.. we will lose selection state on reset here. Perhaps there is a way to avoid this if the ItemsSource implements IKeyIndexMapping, we could look at the selected items still remaining and keep them selected. Can you please file a bug for this on winui ?

---
Refers to: src/Avalonia.Controls/SelectionNode.cs:501 in 8879477. [](commit_id = 88794773229e5c9fe93f29672c41b9cbac601a89, deletion_comment = False)

ranjeshj commented on src/Avalonia.Controls/SelectionModel.cs:
```diff
@@ -36,9 +36,12 @@ public SelectionModel()
             set
             {
-                ClearSelection(resetAnchor: true, raiseSelectionChanged: false);
+                using (var operation = new Operation(this))
```
Would it make sense to expose this ? That way if an app is making multiple changes, they could batch all of them and get a single collection changed event ?

grokys replied:
Very good point!

grokys commented:
> Perhaps there is a way to avoid this if the ItemsSource implements IKeyIndexMapping, we could look at the selected items still remaining and keep them selected. Can you please file a bug for this on winui ?

Yeah I think the only way to retain a selection on reset is to keep a list of the selected items, which `SelectionModel` goes out of its way to avoid... I also think we're going to need this feature in Avalonia :/

 👍 will open an issue to discuss.

grokys committed:
Added some failing tests.

That demonstrate some problems with the `SelectionModel` change notifications found so far.

grokys committed:
Refactor of SelectionModel change notifications.

To address issues found.

grokys commented:
Just pushed some changes to the design of this. 2336f85 adds unit tests for some problems found integrating these changes with real code, and ea968c4 addresses these problems.

A potential problem remains that if the items source is changed during a `SelectionChanged` event and then `SelectedItems`/`DeselectedItems` is read, the items will be incorrect due to these collections being lazily evaluated. Not sure what the best solution for this is. Maybe adding a `CollectionChanged` listener in `SelectionModelChangeSet` which sets a flag causing an exception on any subsequent read of `SelectedItems`/`DeselectedItems`?

Gillibald commented:
Is a SelectionModel bound to the current ItemsSource? So if I would update the ItemsSource the current SelectionModel would change. That way a SelectionModel is always associated with a specific ItemsSource.

So if you are interested in selection changes you also have to listen to changes to the current SelectionModel. Not sure if this would make the above scenario easier to deal with.

ranjeshj commented:
> Is a SelectionModel bound to the current ItemsSource? So if I would update the ItemsSource the current SelectionModel would change. That way a SelectionModel is always associated with a specific ItemsSource.
> 
> So if you are interested in selection changes you also have to listen to changes to the current SelectionModel. Not sure if this would make the above scenario easier to deal with.

If you are providing an ItemsSource to SelectionModel, it will track changes and make sure the same items are selected when indexes change due to collection changes.

grokys commented:
> Is a SelectionModel bound to the current ItemsSource? So if I would update the ItemsSource the current SelectionModel would change. That way a SelectionModel is always associated with a specific ItemsSource.

Yes, it is: `SelectionModel.Source` will be set to `SelectingItemsControl.Items`.

> So if you are interested in selection changes you also have to listen to changes to the current SelectionModel.

Yep, this is how it will work.

> If you are providing an ItemsSource to SelectionModel, it will track changes and make sure the same items are selected when indexes change due to collection changes.

In general, yes, but not in the case of `Reset` currently (as you know @ranjeshj). This is https://github.com/microsoft/microsoft-ui-xaml/issues/1889.

---

However I suspect there may be some misunderstanding around what I was talking about. None of the above relates to the problem I mentioned.

The problem comes about because we don't materialize `SelectionModelSelectionChangedEventArgs.SelectedItems`/`DeselectedItems` until they are accessed. This is because these lists could be _huge_ in the case of a "select all" on a large number of items and may well not be needed.

Perhaps some code will make it clearer:

```csharp
void Foo()
{
	// A source collection with 1 million items.
	selectionModel.Source = Enumerable.Range(0, 1_000_000);
	
	// Listen for selection changes.
	selectionModel.SelectionChanged += OnSelectionChanged;
	
	// Select all 1 million items.
	selectionModel.SelectAll();
}

void OnSelectionChanged(object sender, SelectionModelSelectionChangedEventArgs e)
{
	// At this point SelectionModelSelectionChangedEventArgs simply holds the information
	// that items [0..1,000,000] have been selected. e.SelectedItems, e.SelectedIndexes
	// are not initialized.
	
	// Now we set a new source on selectionModel with only 10 items.
	selectionModel.Source = Enumerable.Range(0, 10);
	
	// Reading SelectedItems now is going to fail, because it's going to try to look up
	// items [0..1,000,000] in a source collection of only 10 items.
	var selected = e.SelectedItems;
}
```

Note that in the case above, the failure will be obvious because there are no longer enough items. But this could also return the _wrong_ items (instead of failing) if the source is mutated, which is the real danger here.

grokys committed:
Lazily create the selected/deselected lists.

Gillibald commented:
I mean that there should be one SelectionModel per ItemsSource instance. So if you would change the ItemsSource you would get a new SelectionModel.

grokys commented:
Actually I think if I use `SelectedItems` to return the selected/deselected items as mentioned in https://github.com/AvaloniaUI/Avalonia/pull/3470#discussion_r373383235 then that should automatically handle this (though I suspect I've not ported that part correctly and there's no unit tests for that functionality).

grokys commented:
> I mean that there should be one SelectionModel per ItemsSource instance. So if you would change the ItemsSource you would get a new SelectionModel.

Why is this? `SelectionModel` has a `Source` property that can be changed. Is this related to what I was talking about or are you talking about something different?

Gillibald commented:
Related to the issue with changing the source during the selection change event. By having a SelectionMode per ItemsSource you don't have this issue. So in the end source never changes on the SelectionModel.

grokys commented:
@Gillibald the problem is with the items _in_ the collection changing, not the `Items` property being reassigned.

Gillibald commented:
Ahh okay I may have misread the "items source is changed during a SelectionChanged event"

Now I get the picture 👍

grokys committed:
Use SelectedItems for change event args.

grokys commented:
Actually I've taken a look at how WinUI handles this is `SelectionModel.SelectedIndices`/`SelectedItems` and it doesn't handle it there either. This test in WinUI:

```
var target = new SelectionModel();
var data = new ObservableCollection<string> { "foo", "bar", "baz", "qux" };

target.Source = data;
target.SelectRange(IndexPath.CreateFrom(1), IndexPath.CreateFrom(2));

foreach (var i in target.SelectedIndices)
{
	data.RemoveAt(0);
}
```

Returns `R.1` twice. Not sure it's worth worrying about TBH.

grokys committed:
Add nullable annotations.

grokys committed:
Handle null SelectionModel.Source.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Expose API for batch updates.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Added ISelectionModel.Update().

grokys committed:
Added failing test.

Clearing a nested selection doesn't raise `SelectionChanged`.

grokys committed:
Fix clearing nested selection not raising SelectionChanged.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Fix SelectionNode.Cleanup.

- Removed disposed child nodes
- Don't dispose child node with descendent selection

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys commented:
Merged as part of https://github.com/AvaloniaUI/Avalonia/pull/3583
